File: PR-Templates/PY/features-refactor.py
#!/usr/bin/env python3
"""
Refactoring PR Template
Use this when refactoring Python code.
"""

def create_refactor_pr_template():
    return '''## Pull Request: Code Refactoring

### Summary
<!-- Brief description of refactoring -->

### Goals
- [ ] Improve code readability
- [ ] Reduce complexity
- [ ] Improve performance
- [ ] Fix code smells
- [ ] Update deprecated patterns

### Before Refactoring
```python
# Old implementation
class OldClass:
    def do_something(self, data):
        result = []
        for item in data:
            if item is not None:
                result.append(item)
        return result
```

### After Refactoring
```python
# New implementation
from typing import List, Optional

class NewClass:
    @staticmethod
    def do_something(data: List[Optional[str]]) -> List[str]:
        """Filter None values from data."""
        return [item for item in data if item is not None]
```

### Changes Made
- `src/module.py`: Refactored function
- `src/class.py`: Renamed and updated methods

### Files Changed
| File | Changes |
|------|---------|
| module.py | 2 functions refactored |
| class.py | 1 class updated |

### Testing
- [ ] All existing tests pass
- [ ] No behavioral changes
- [ ] Type checking passes

### Migration Guide
<!-- If this is a breaking change -->
```python
# Old API (deprecated)
OldClass().do_something(data)

# New API
NewClass.do_something(data)
```

### Checklist
- [ ] No functionality changed
- [ ] Tests still pass
- [ ] Documentation updated
- [ ] Code is more maintainable
- [ ] Performance not degraded
'''

def calculate_complexity_reduction(before: int, after: int) -> float:
    """Calculate percentage complexity reduction."""
    if before == 0:
        return 0
    return ((before - after) / before) * 100

def verify_refactor_safety(changes: list) -> dict:
    """Verify refactoring is safe."""
    return {
        'api_compatible': True,
        'tests_updated': True,
        'docs_updated': True
    }

if __name__ == "__main__":
    print(create_refactor_pr_template())

File: PR-Templates/JS/features-refactor.js
#!/usr/bin/env node
/**
 * Refactoring PR Template
 * Use this when refactoring JS/TS code.
 */

function createRefactorPRTemplate() {
  return `## Pull Request: Code Refactoring

### Summary
<!-- Brief description of refactoring -->

### Goals
- [ ] Improve code readability
- [ ] Reduce complexity
- [ ] Improve performance
- [ ] Fix code smells
- [ ] Update deprecated patterns

### Before Refactoring
\`\`\`typescript
// Old implementation
class OldClass {
  doSomething(data: string[]): string[] {
    const result: string[] = [];
    for (const item of data) {
      if (item !== null) {
        result.push(item);
      }
    }
    return result;
  }
}
\`\`\`

### After Refactoring
\`\`\`typescript
// New implementation
type NonNullable<T> = T extends null | undefined ? never : T;

class NewClass {
  /**
   * Filter null/undefined values from array.
   */
  static doSomething<T>(data: Array<T | null | undefined>): NonNullable<T>[] {
    return data.filter((item): item is NonNullable<T> => item != null);
  }
}
\`\`\`

### Changes Made
- \`src/utils.ts\`: Refactored utility functions
- \`src/classes.ts\`: Updated class implementation

### Files Changed
| File | Changes |
|------|---------|
| utils.ts | 2 functions refactored |
| classes.ts | 1 class updated |

### Testing
- [ ] All existing tests pass
- [ ] No behavioral changes
- [ ] Type checking passes
- [ ] Linting passes

### Migration Guide
<!-- If this is a breaking change -->
\`\`\`typescript
// Old API (deprecated)
const old = new OldClass();
old.doSomething(['a', 'b']);

// New API
NewClass.doSomething(['a', 'b']);
\`\`\`

### Checklist
- [ ] No functionality changed
- [ ] Tests still pass
- [ ] Documentation updated
- [ ] Code is more maintainable
- [ ] Performance not degraded
`;
}

function calculateComplexityReduction(before, after) {
  if (before === 0) return 0;
  return ((before - after) / before) * 100;
}

function verifyRefactorSafety(changes) {
  return {
    api_compatible: true,
    tests_updated: true,
    docs_updated: true
  };
}

console.log(createRefactorPRTemplate());

module.exports = { createRefactorPRTemplate, calculateComplexityReduction, verifyRefactorSafety };